services/payments/service.go

```diff
@@ -178,3 +191,338 @@ func (s *Service) DecryptBootstrap(ctx context.Context, ciphertext []byte) (map[
+func validateTransactionHistory(
+	ctx context.Context,
+	idempotencyKey *uuid.UUID,
+	transactionHistory []QLDBPaymentTransitionHistoryEntry,
+	kmsClient wrappedKMSClient,
+) (bool, error) {
+	var (
+		reason error
+		err    error
+	)
+	for i, transaction := range transactionHistory {
+		var transactionData Transaction
+		err = ion.Unmarshal(transaction.Data.Data, &transactionData)
+		if err != nil {
+			return false, fmt.Errorf("failed to unmarshal transaction data: %w", err)
+		}
+
+		// Before starting State validation, check that keys and signatures for the record are valid.
+
+		// GenerateIdempotencyKey will verify that the ID is internally consistent within the Transaction.
+		dataIdempotencyKey, err := transactionData.GenerateIdempotencyKey(ctx)
+		if err != nil {
+			return false, fmt.Errorf("ID invalid: %w", err)
+		}
+		// The data object is serialized in QLDB, but when deserialized should contain an ID that matches
+		// the ID on the top level of the QLDB record
+		if *dataIdempotencyKey != *idempotencyKey {
+			return false, fmt.Errorf("top level ID does not match Transaction ID: %s, %s", dataIdempotencyKey, idempotencyKey)
+		}
+		// Each transaction's signature must be verified
+		txID := transaction.Data.IdempotencyKey.String()
+		verifyOutput, err := kmsClient.Verify(ctx, &kms.VerifyInput{
+			KeyId:   &txID,
+			Message: transaction.Data.Data,
+		})
+		if err != nil {
+			return false, fmt.Errorf("failed to verify signature: %w", err)
+		}
+		if !verifyOutput.SignatureValid {
+			return false, fmt.Errorf("signature for record %s invalid with metadata: %v", transaction.Data.IdempotencyKey.String(), verifyOutput.ResultMetadata)
+		}
+
+		// Now that the data itself is verified, proceed to check transition States.
+		transactionState := transactionData.State
+		// Transitions must always start at 0
+		if i == 0 {
+			if transactionState != 0 {
+				return false, errors.New("initial state is not valid")
+			}
+			continue
+		}
+		var previousTransitionData Transaction
+		err = ion.Unmarshal(transactionHistory[i-1].Data.Data, &previousTransitionData)
+		if err != nil {
+			return false, fmt.Errorf("failed to unmarshal previous transition history record: %w", err)
+		}
+		previousIdempotencyKey, err := previousTransitionData.GenerateIdempotencyKey(ctx)
+		if err != nil {
+			return false, fmt.Errorf("ID invalid: %w", err)
+		}
+		// The IdempotencyKeys of all records in the transition history of a Transaction should match
+		if *dataIdempotencyKey != *previousIdempotencyKey {
+			return false, fmt.Errorf("idempotencyKeys in transition history do not match: %s, %s", idempotencyKey.String(), previousIdempotencyKey.String())
+		}
+		previousTransitionState := previousTransitionData.State
+		// New transaction state should be present in the list of valid next states for the "previous" (current) state.
+		if !slices.Contains(Transitions[previousTransitionState], transactionState) {
```

Comment:
this is slightly different then the logic in `nextStateValid`, will that be an issue?

Reply:
I should just be using `nextStateValid` here. The code as written works as expected, but there's no reason to duplicate the underlying check and risk future issues. Fixed.